### PR TITLE
This PR is initial support for targetid.  It includes significant fix…

### PR DESCRIPTION
**Pull request by gregrodgers**
…es for multiple GPU compilation which is a merge of upstream architecture for multiple GPUs with OpenMP.   That architecture uses one INSTANCE of a toolchain for each GPU.  This PR also includes end-to-end fixes for the active-offload-target utility and the openmp runtime to check active target with target info stored in the image info datastructures now created by clang-offload-wrapper.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp [hunk 1]
```diff
@@ -356,29 +356,61 @@ const char *AMDGCN::OpenMPLinker::constructLlcCommand(
   return LlcOutputFile;
 }
 
+//  Similar to AMDGCN::Linker::constructLldCommand in HIP.cpp
+//  FIXME: Add Ron's change to support GetEnv("ROCM_LLD_ARGS")
```
Comment (ronlieb) [inline clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:360]: or make Ron do it later.

**Comment by gregrodgers** [comment]
I realize the update to Driver.cpp is big.  However, it is a significantly smaller diff to upstream Driver.cpp because much of the OpenMPAction Builder was taken from upstream code.   So the old aomp OpenMPAction Builder is completely gone.

The upstream OpenMPAction Builder uses one offload ToolChain INSTANCE per offload GPU whereas it was a single offload Toolchain in the past.  Note that Cuda and HIP use a single offload toolchain. 

This update also starts to deprecate the use of -Xopenmp-target=<triple> -march=  command line syntax.  The GPU with targetid features can now be specified as the environment string in the triple of the -fopenmp-targets option.  This is a significant simplification of command line syntax.   

We still need to add updates to the Driver toolchains to support heterogenous library unbundling.   As I said, this is "initial" support.

**Review by ronlieb — COMMENTED** [review]
what are some examples of how the command line will replace the old -X options?
do  you have a few tests ready to add to perhaps test/omp5.0 ?
Do you have an example of how 
 #pragma omp requires unified_shared_memory 
should/will interact with target-id ?

**Comment by gregrodgers** [comment]
This is old method (which still works) 
clang -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march gfx906  foo.cpp
This is new method 
clang -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa-gfx906  foo.cpp

**Comment by gregrodgers** [comment]
Maybe should automate the -fopenmp when -fopenmp-targets is set.  Maybe we should make the triple optional for targets starting with gfx or sm_.    This would make the above example be . 

clang -fopenmp-targets=gfx906   foo.cpp 

This is a good time to make this feature/change.

**Comment by ronlieb** [comment]
[AMD Official Use Only - Internal Distribution Only]

Both awesome simplifications.

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -2231,6 +2261,21 @@ void Driver::BuildInputs(const ToolChain &TC, DerivedArgList &Args,
 
         // stdin must be handled specially.
         if (memcmp(Value, "-", 2) == 0) {
+          if (IsFlangMode()) {
```
Comment (yxsamliu) [inline clang/lib/Driver/Driver.cpp:2264]: this change is irrelevant to target ID. should be separated to another patch

File: clang/lib/Driver/Driver.cpp [hunk 3]
```diff
@@ -3001,10 +3046,9 @@ class OffloadingActionBuilder final {
       }
 
       // By default, we produce an action for each device arch.
-      for (Action *&A : CudaDeviceActions) {
```
Comment (yxsamliu) [inline clang/lib/Driver/Driver.cpp:3004]: formating change should be removed

File: clang/lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -3026,8 +3070,8 @@ class OffloadingActionBuilder final {
         // Linking all inputs for the current GPU arch.
         // LI contains all the inputs for the linker.
         OffloadAction::DeviceDependences DeviceLinkDeps;
-        DeviceLinkDeps.add(*DeviceLinkAction, *ToolChains[0],
```
Comment (yxsamliu) [inline clang/lib/Driver/Driver.cpp:3029]: formating change should be removed

**Comment by JonChesterfield** [comment]
This is too big a change to meaningfully review. Reading through it I see old code commented out and printf debugging, but going through and picking out the nits seems unproductive. Are you able to split this into the nonfunctional change and the semantically orthogonal changes?

In particular, it would be really useful for the upstream porting effort to first land the part that makes this more like trunk.

**Comment by gregrodgers** [comment]
This PR is intended to get targetid working in AOMP and to discuss the architecture for doing it.   Organizing the upstream efforts for this is a later task.   We need to be sure we can make this work end-to-end first.  We know there are major changes before upstreaming this.  In fact, it is debatable if we can use environment field in the triple to track the targetid.   But this PR shows how that can be done.  We need to look for situations where using the triple for the targetid would break stuff.      

I did ask Singh to arrange for the fix in Action.cpp to be upstreamed first.   The code in Driver.cpp absolutely needs this fix.   

Also, I never expected the bash script active-offload-target to be upstreamable.  This needs to be rewritten as a c++ tool after we settle on the user interface.

**Comment by ronlieb** [comment]
i  see a lot of code came in from the HIP AMDGCN::Linker::constructLldCommand implimentation.
i guess its a good thing, something to watch out for later if you see any issues.

**Comment by saiislam** [comment]
Following clang lit tests are failing after my patch:

1.   Clang :: Driver/amdgcn-openmp-toolchain-dwarf.c
2.   Clang :: Driver/amdgcn-toolchain-openmp-duplicate-arguments.c
3.   Clang :: OpenMP/driver-openmp-target.c

Taking a look.

**Comment by gregrodgers** [comment]
we need to delete this pr and branch and rebase it to next aomp-dev branch
